components/scream/src/physics/spa/spa_functions_impl.hpp

```diff
@@ -0,0 +1,285 @@
+        ccn3_src_sub(k) = linear_interp(ccn3_beg_sub(k),ccn3_end_sub(k),t_norm);
+    });
+    team.team_barrier();
+    }
+    {
+    /* Loop over all SW variables with nswbands */
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, nswbands), [&] (int n) {
+      const auto& aer_g_sw_beg_sub           = ekat::subview(data_beg.AER_G_SW, i, n);
+      const auto& aer_g_sw_end_sub           = ekat::subview(data_end.AER_G_SW, i, n);
+      const auto& aer_g_sw_src_sub           = ekat::subview(aer_g_sw_src, i, n);
+
+      const auto& aer_ssa_sw_beg_sub         = ekat::subview(data_beg.AER_SSA_SW, i, n);
+      const auto& aer_ssa_sw_end_sub         = ekat::subview(data_end.AER_SSA_SW, i, n);
+      const auto& aer_ssa_sw_src_sub         = ekat::subview(aer_ssa_sw_src, i, n);
+  
+      const auto& aer_tau_sw_beg_sub         = ekat::subview(data_beg.AER_TAU_SW, i, n);
+      const auto& aer_tau_sw_end_sub         = ekat::subview(data_end.AER_TAU_SW, i, n);
+      const auto& aer_tau_sw_src_sub         = ekat::subview(aer_tau_sw_src, i, n);
+
+      /* Now loop over fastest index, the number of vertical packs */
+      Kokkos::parallel_for(
+        Kokkos::ThreadVectorRange(team, nk_pack), [&] (Int k) {
+        aer_g_sw_src_sub(k)   = linear_interp(aer_g_sw_beg_sub(k),   aer_g_sw_end_sub(k),   t_norm);
+        aer_ssa_sw_src_sub(k) = linear_interp(aer_ssa_sw_beg_sub(k), aer_ssa_sw_end_sub(k), t_norm);
+        aer_tau_sw_src_sub(k) = linear_interp(aer_tau_sw_beg_sub(k), aer_tau_sw_end_sub(k), t_norm);
+      });
+    });
+    team.team_barrier();
+    }
+    {
+    /* Loop over all LW variables with nlwbands */
+    Kokkos::parallel_for(
+      Kokkos::TeamThreadRange(team, nlwbands), [&] (int n) {
+      const auto& aer_tau_lw_beg_sub         = ekat::subview(data_beg.AER_TAU_LW, i, n);
+      const auto& aer_tau_lw_end_sub         = ekat::subview(data_end.AER_TAU_LW, i, n);
+      const auto& aer_tau_lw_src_sub         = ekat::subview(aer_tau_lw_src, i, n);
+
+      /* Now loop over fastest index, the number of vertical packs */
+      Kokkos::parallel_for(
+        Kokkos::ThreadVectorRange(team, nk_pack), [&] (Int k) {
+        aer_tau_lw_src_sub(k) = linear_interp(aer_tau_lw_beg_sub(k), aer_tau_lw_end_sub(k), t_norm);
+      });
+    });
+    team.team_barrier();
+    }
+  });
+  Kokkos::fence();
+
+  // Third Step: Vertical interpolation, project the SPA data onto the pressure profile for this simulation.
+  // This is done using the EKAT linear interpolation routine, see /externals/ekat/util/ekat_lin_interp.hpp
+  // for more details. 
+  using LIV = ekat::LinInterp<Real,Spack::n>;
+  Real minthreshold = 0.0;  // Hard-code a minimum value for aerosol concentration to zero.
+
+  LIV VertInterp(ncols_atm,pressure_state.nlevs,nlevs_atm,minthreshold);
+  /* Parallel loop strategy:
+   * 1. Loop over all simulation columns (i index)
+   * 2. Where applicable, loop over all aerosol bands (n index)
+   */ 
```

Comment:
You'll need to use a custom policy.
```
const Int most_bands = std::max(nlwbands, nswbands);
LIV::TeamPolicy policy(ncols_atm, ekat::OnGpu<LIV::ExeSpace>::value ? most_bands: 1, VertInterp.km2_pack());
```

Reply:
@bartgol , can you confirm this is the correct policy?

Reply:
I think it is correct.

Reply:
@jgfouca  

I find it annoying to have to do the switch based on OnGpu<ExeSpace> though. In Hommexx, we handled this from our `get_default_team_policy` util, which used a ThreadPreferences input struct to specify the max team size (usually the length of the 2nd nested for) and the max vector size (the length of the 3rd nested loop), whether to prefer threads over vectors, and whether to prefer large teams even if the outer ||for had several iters. The routine then, based on thread preferences and arch, would spit out the team policy you needed.

We might add something like that to ekat, at some point. You can find what I'm talking about in `components/homme/src/share/cxx/ExecSpaceDefs.*pp`.